src/setup_payload/tests/TestAdditionalDataPayload.cpp

```diff
@@ -0,0 +1,197 @@
+                                                                         bufferHandle, additionalDataFields);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+
+    char output[51];
+    err = BytesToUppercaseHexString(bufferHandle->Start(), bufferHandle->DataLength(), output, ArraySize(output));
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(output, "5500003000120A00D1414E39C9A08469DA2E29F0ACAA086618") == 0);
+
+    AdditionalDataPayload resultPayload;
+    err = AdditionalDataPayloadParser(bufferHandle->Start(), bufferHandle->DataLength()).populatePayload(resultPayload);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(resultPayload.rotatingDeviceId.c_str(), "0A00D1414E39C9A08469DA2E29F0ACAA0866") == 0);
+}
+
+void TestGeneratingAdditionalDataPayloadWithoutRotatingIdAndParsingIt(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferHandle bufferHandle;
+
+    char serialNumber[]      = "89051AAZZ236";
+    uint16_t lifetimeCounter = 10;
+    BitFlags<AdditionalDataFields> additionalDataFields;
+
+    // Init Chip memory management before the stack
+    chip::Platform::MemoryInit();
+
+    err = AdditionalDataPayloadGenerator().generateAdditionalDataPayload(lifetimeCounter, serialNumber, ArraySize(serialNumber),
+                                                                         bufferHandle, additionalDataFields);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+
+    char output[51];
+    err = BytesToUppercaseHexString(bufferHandle->Start(), bufferHandle->DataLength(), output, ArraySize(output));
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(output, "55000018") == 0);
+
+    AdditionalDataPayload resultPayload;
+    err = AdditionalDataPayloadParser(bufferHandle->Start(), bufferHandle->DataLength()).populatePayload(resultPayload);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(resultPayload.rotatingDeviceId.c_str(), "") == 0);
+}
+
+/**
+ *  Test Suite that lists all the Test functions.
+ */
+// clang-format off
+const nlTest sTests[] =
+{
+    NL_TEST_DEF("Test Generating Additional Data Payload with RotatingId", TestGeneratingAdditionalDataPayloadWithRotatingId),
+    NL_TEST_DEF("Test Generating Additional Data Payload without RotatingId", TestGeneratingAdditionalDataPayloadWithoutRotatingId),
+    NL_TEST_DEF("Test Generating Additional Data Payload with RotatingId + Parsing it", TestGeneratingAdditionalDataPayloadWithRotatingIdAndParsingIt),
+    NL_TEST_DEF("Test Generating Additional Data Payload without RotatingId + Parsing it", TestGeneratingAdditionalDataPayloadWithoutRotatingIdAndParsingIt),
```

Comment:
Can you add a few tests for generating and parsing the string format?

Reply:
Parsing is implemented only for the additional payload which have the binary representation, we treat the rotating id as an opaque value, however i added some code for parsing it "partially", since the first part is the lifetime counter, the second part (which is the hash suffix) cannot be parsed since it's part of the hash of multiple inputs.